File: algorithms/125.验证回文串.js
/*
 * @lc app=leetcode.cn id=125 lang=javascript
 *
 * [125] 验证回文串
 *
 * https://leetcode-cn.com/problems/valid-palindrome/description/
 *
 * algorithms
 * Easy (40.65%)
 * Likes:    101
 * Dislikes: 0
 * Total Accepted:    47.2K
 * Total Submissions: 116K
 * Testcase Example:  '"A man, a plan, a canal: Panama"'
 *
 * 给定一个字符串，验证它是否是回文串，只考虑字母和数字字符，可以忽略字母的大小写。
 *
 * 说明：本题中，我们将空字符串定义为有效的回文串。
 *
 * 示例 1:
 *
 * 输入: "A man, a plan, a canal: Panama"
 * 输出: true
 *
 *
 * 示例 2:
 *
 * 输入: "race a car"
 * 输出: false
 *
 *
 */
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
  s = s.toLowerCase();
  s = s.replace(/\W/g, "");
  let lp = -1,
    rp = s.length;
  while (++lp < --rp) {
    if (s[lp] !== s[rp]) {
      return false;
    }
  }
  return true;
};

File: algorithms/golang/125.验证回文串.go
/*
 * @lc app=leetcode.cn id=125 lang=golang
 *
 * [125] 验证回文串
 */

// @lc code=start
func isUpperCase(ch byte) bool {
	if ch >= 'A' && ch <= 'Z' {
		return true
	}
	return false
}
func isLowerCase(ch byte) bool {
	if ch >= 'a' && ch <= 'z' {
		return true
	}
	return false
}
func isNumber(ch byte) bool {
	if ch >= '0' && ch <= '9' {
		return true
	}
	return false
}
func isValidLetter(ch byte) bool {
	if isLowerCase(ch) || isUpperCase(ch) || isNumber(ch) {
		return true
	}
	return false
}
func toLowerCase(ch byte) byte {
	if isUpperCase(ch) {
		return ch + 'a' - 'A'
	}
	return ch
}
func isPalindrome(s string) bool {
	for l, r := 0, len(s)-1; l < r; {
		ch_l, ch_r := s[l], s[r]
		for l < r && !isValidLetter(ch_l) {
			l++
			ch_l = s[l]
		}
		for r > l && !isValidLetter(ch_r) {
			r--
			ch_r = s[r]
		}
		if l < r && toLowerCase(ch_l) != toLowerCase(ch_r) {
			return false
		}
		l++
		r--
	}
	return true
}

// @lc code=end
